threedworld-mit/tdw
# get_point_cloud

Consolidates point cloud generation code into a TDWUtils function.

The original code:

<details>

```python
import argparse
import json
import os.path as osp
import time

import cv2
import numpy as np


Image_changed = True
Height = None
Width = None
Iinv = None
Cam_to_img_mat = None
Img_pixs_ones = None


def img_to_world(depth, rgb, E, depth_threshold=None):
    global Image_changed, Height, Width, Iinv, Cam_to_img_mat, Img_pixs_ones
    # use global variable here to compute some values just once
    # so as to speed up calculation
    if Height is None or Height != depth.shape[0]:
        Height = depth.shape[0]
        Image_changed = True
    if Width is None or Width != depth.shape[1]:
        Width = depth.shape[1]
        Image_changed = True
    if Image_changed:
        img_pixs = np.mgrid[0: depth.shape[0], 0: depth.shape[1]].reshape(2, -1)
        img_pixs[[0, 1], :] = img_pixs[[1, 0], :]  # swap (v, u) into (u, v)
        Img_pixs_ones = np.concatenate((img_pixs, np.ones((1, img_pixs.shape[1]))))
        if Iinv is None:
            I = get_intrinsics(height=Height, width=Width, vfov=54.43222)
            Iinv = np.linalg.inv(I[:3, :3])
        Cam_to_img_mat = np.dot(Iinv, Img_pixs_ones)
        #print(Cam_to_img_mat.shape, Cam_to_img_mat)
        Image_changed = False

    depth_vals = depth.reshape(-1)
    #rgb_vals = rgb.reshape(-1, 3)
    rgb_vals = None
    if rgb is not None:
        assert type(depth_threshold) is tuple
        valid = depth_vals > depth_threshold[0]
        if len(depth_threshold) > 1:
            valid = np.logical_and(valid,
                                   depth_vals < depth_threshold[1])
        depth_vals = depth_vals[valid]
        rgb_vals = rgb_vals[valid]
        points_in_cam = np.multiply(Cam_to_img_mat[:, valid], depth_vals)
    else:
        points_in_cam = np.multiply(Cam_to_img_mat, depth_vals)
    points_in_cam = np.concatenate((points_in_cam,
                                    np.ones((1, points_in_cam.shape[1]))),
                                   axis=0)
    #return points_in_cam
    #print(points_in_cam.shape, points_in_cam)
    #print(points_in_cam[:, int(points_in_cam.shape[1] / 2)])
    points_in_world = np.dot(E, points_in_cam)
    #points_in_world = points_in_world.T
    return points_in_world, rgb_vals


def get_intrinsics(height, width, vfov):
    # calculate the intrinsic matrix from vertical_fov
    # notice that hfov and vfov are different if height != width
    # we can also get the intrinsic matrix from opengl's
    # perspective matrix
    # http://kgeorge.github.io/2014/03/08/calculating-opengl-
    # perspective-matrix-from-opencv-intrinsic-matrix
    vfov = vfov / 180.0 * np.pi
    tan_half_vfov = np.tan(vfov / 2.0)
    tan_half_hfov = tan_half_vfov * width / float(height)
    fx = width / 2.0 / tan_half_hfov  # focal length in pixel space
    fy = height / 2.0 / tan_half_vfov
    I = np.array([[fx, 0, width / 2.0],
                  [0, fy, height / 2.0],
                  [0, 0, 1]])
    return I


def gen_point_cloud(depth, rgb, E, depth_threshold=[0.0, 100.0], inv_E=True):
    if inv_E:
        E = np.linalg.inv(np.array(E).reshape((4, 4)))
    # different from real-world camera coordinate system
    # opengl uses negative z axis as the camera front direction
    # x axes are same, hence y axis is reversed as well
    # https://learnopengl.com/Getting-started/Camera
    rot = np.array([[1, 0, 0, 0],
                    [0, -1, 0, 0],
                    [0, 0, -1, 0],
                    [0, 0, 0, 1]])
    E = np.dot(E, rot)
    points, _ = img_to_world(depth, rgb, E)
    return points[:3, :], None


def show_point_cloud(color_path, depth_path, extrinsics):
    depth = cv2.imread(depth_path).astype(np.float32)
    depth = (depth[:, :, 2] * 256.0 * 256.0 + depth[:, :, 1] * 256.0 + depth[:, :, 0]) / (256 ** 3) * 100.1

    #depth = depth.astype(np.float32) / 255.0 * 100.1
    rgb = cv2.imread(color_path)
    rgb = rgb[:, :, ::-1]
    #print(depth)
    return gen_point_cloud(depth, rgb, extrinsics)

Image_changed = True
Height = None
Width = None
Iinv = None


id = '657'
depth = np.load(f'./sma_depth/depth_editor/depth{id}.npy')
E = np.load(f'./sma_depth/depth_editor/camera_matrix{id}.npy')
pc, _ = gen_point_cloud(depth, rgb=None, E=E)
#pc = img_to_world(depth, rgb=None, E=E)
print(pc.shape)
pc = pc.reshape(3, 256, 256)
f = open(f'gpc{id}.txt', 'w')
for i in range(pc.shape[1]):
    for j in range(pc.shape[2]):
        if pc[2, i, j] < 99:
            f.write(f'{pc[0, i, j]} {pc[1, i, j]} {pc[2, i, j]}\n')

```

</details>

# How to Test

1. Make sure that the `tdw` module is installed locally. If you installed it via `pip3 install -e .` you don't need to do anything. If you installed it via `pip3 install tdw` you need to do this: `pip3 uninstall tdw` then `cd path/to/tdw` (replace `path/to` with the actual path) then `cd Python` then `pip3 install -e .`
2. `git fetch` and `git checkout point_cloud`
3. Download these files: [point_cloud.zip](https://github.com/threedworld-mit/tdw/files/5736455/point_cloud.zip)
4. Run this code:

```python
from pathlib import Path
import numpy as np
from tdw.tdw_utils import TDWUtils


p = Path.home().joinpath("00000317_depth.png")
image = p.read_bytes()
depth_values = TDWUtils.get_depth_values(image=image)
dest = Path.home().joinpath("point_cloud_tdwutils.txt")
camera_matrix_path = Path.home().joinpath("camera_matrix317.npy")
camera_matrix = np.load(str(camera_matrix_path.resolve()))
TDWUtils.get_point_cloud(depth=depth_values, filename=str(dest.resolve()), camera_matrix=camera_matrix)
```

5. Import the data to Meshlab.

# Changelog

### `tdw` module

#### `TDWUtils`

- Addded: `get_point_cloud()` Create a point cloud from an numpy array of depth values.
